pkg/fqdn/cache.go

```diff
@@ -640,3 +640,220 @@ func (c *DNSCache) UnmarshalJSON(raw []byte) error {
 
 	return nil
 }
+
+// DNSPendingDelete is an IP that has expired or been evicted from a DNS cache.
+// It records the DNS name and IP, along with other bookkeeping timestamps that
+// help determine when it can be finally deleted. Entries are deletable when
+// they are not marked active by CT GC.
+type DNSPendingDelete struct {
+	Names           []string  `json:"names,omitempty"`
+	IP              net.IP    `json:"ip,omitempty"`
+	ActiveAt        time.Time `json:"active-at,omitempty"`
+	DeletePendingAt time.Time `json:"delete-pending-at,omitempty"`
+}
+
+// DNSPendingDeletes collects DNS Name->IP mappings that may be deletable. They
+// are marked active, periodically, by the CT GC goroutine. When .GC is called,
+// active and deletable entries are returned, allowing us to skip deleting an IP
+// from the global DNS cache to avoid breaking connections that outlast the DNS
+// TTL.
+type DNSPendingDeletes struct {
+	lock.Mutex
+	deletes        map[string]*DNSPendingDelete // map[ip]toDelete
+	lastCTGCUpdate time.Time
+}
+
+// NewDNSPendingDeletes constructs a DNSPendingDeletes that is read to use
+func NewDNSPendingDeletes() *DNSPendingDeletes {
+	return &DNSPendingDeletes{
+		deletes: make(map[string]*DNSPendingDelete),
+	}
+}
+
+// WantToDelete enqueues the ip -> qname as a possible deletion
+// updatedExisting is true when an earlier enqueue existed and was updated
+func (d *DNSPendingDeletes) WantToDelete(ipStr string, qname ...string) (updatedExisting bool) {
+	d.Lock()
+	defer d.Unlock()
+
+	entry, updatedExisting := d.deletes[ipStr]
+	if !updatedExisting {
+		entry = &DNSPendingDelete{}
+		d.deletes[ipStr] = entry
+	}
+
+	entry.Names = append(entry.Names, qname...)
+	entry.IP = net.ParseIP(ipStr)
+	entry.DeletePendingAt = time.Now()
+
+	return updatedExisting
+}
+
+// Delete after it is seen by CT GC, DeletePendingAt < lastCTGCUpdate, and the ActiveTime is 0 or no longer current, ActiveAt < lastCTGCUpdate
+func (d *DNSPendingDeletes) canDelete(del *DNSPendingDelete) bool {
+	return d.lastCTGCUpdate.After(del.DeletePendingAt) && d.lastCTGCUpdate.After(del.ActiveAt)
+}
+
+// GC returns active and deletable DNSPendingDelete entries. This removes
+// deletable entries interally, and repeated calls will return different data.
+// Entries are active if they have been marked active (with MarkActive). When
+// SetCTGCTime is called and an entry not marked active, it become deletable.
+// Calling WantToDelete on a deletable entry will make it active again.
+func (d *DNSPendingDeletes) GC() (active, deletable []*DNSPendingDelete) {
+	d.Lock()
+	defer d.Unlock()
+
+	// Collect entries we can delete
+perIP:
+	for _, del := range d.deletes {
+		cpy := &DNSPendingDelete{
```

Comment:
Why don't we setup a target to create a deepcopy method of this structure instead?